openrarity/collection/collection.py

```diff
@@ -0,0 +1,337 @@
+        self._input_checksum: str = self._hash_data(cast(JsonEncodable, tokens))
+        self._ranks_checksum: str | None = None
+        self._token_supply: int | dict[str | int, int]
+        (self._token_supply, self._tokens) = validate_tokens(token_type, tokens)
+
+        # Derived data
+        self._vertical_attribute_data: list[ValidatedTokenAttribute] = []
+        self._attribute_statistics: AttributesStatistics = {
+            "string": [],
+            "number": [],
+            "date": [],
+        }
+        self._entropy: float | None = None
+        self._token_statistics: list[TokenStatistic] = []
+        self._string_types: list[ValidatedTokenAttribute] = []
+        self._number_types: list[ValidatedTokenAttribute] = []
+        self._date_types: list[ValidatedTokenAttribute] = []
+
+        # Output data
+        self._ranks: list[RankedToken] = []
+
+    def __repr__(self) -> str:
+        return f"Collection({self._token_type})"
+
+    @property
+    def tokens(self) -> dict[TokenId, RawToken]:
+        return self._tokens
+
+    @property
+    def total_supply(self) -> int:
+        # SemiFungible needs to sum the supply of each token
+        if isinstance(self._token_supply, dict):
+            return sum(self._token_supply.values())
+
+        return self._token_supply
+
+    @property
+    def attribute_statistics(self) -> AttributesStatistics:
+        if not self._attribute_statistics:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._attribute_statistics
+
+    @property
+    def entropy(self) -> float | None:
+        return self._entropy
+
+    @property
+    def token_statistics(self) -> list[TokenStatistic]:
+        if not self._token_statistics:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._token_statistics
+
+    @property
+    def ranks(self) -> list[RankedToken]:
+        if not self._ranks:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._ranks
+
+    @property
+    def checksum(self) -> str:
+        if not self._ranks_checksum:
+            raise AttributeError(
+                f"Please run '{repr(self)}.rank_collection()' to view this property"
+            )
+        return self._input_checksum + self._ranks_checksum
+
+    @overload
+    def rank_collection(  # type: ignore
+        self,
+        rank_by: tuple[
+            Literal[
+                "metric.probability",
+                "metric.information",
+                "metric.information_entropy",
+                "metric.unique_trait_count",
+                "metric.max_trait_information",
+            ],
+            ...,
+        ] = ("metric.unique_trait_count", "metric.information_entropy"),
```

Comment:
that assignment looks a bit cryptic to me. Should We simplify this by bringing the type to some predefined static variable something like OR_SORT_FUNCTION =("metric.unique_trait_count", "metric.information_entropy")

Reply:
The `...` in `tuple[]` just means that it can be of arbitrary length. So specifically the above typing says that it is a tuple with any number of the above literal values.

Reply:
I am indifferent. Its the openrarity library and a default value was provided so it can be assumed that it is the OR preferred one. Masking it with a static variable will mess up introspection I believe.